Added slides on C++20 ranges (#242)
hsf-training/cpluspluscourse: talk/morelanguage/ranges.tex

\subsection[range]{Ranges}

\begin{frame}[fragile]
  \frametitlecpp[20]{Ranges}
  \begin{block}{The range concept}
    \begin{itemize}

>     \item A range is a group of items that you can iterator over

Reviewer: ```suggestion
    \item A range is a group of items that you can iterate over
```

    \item It must provide iterators to its beginning and end, obtainable via \mintinline{cpp}{std::range::begin} and \mintinline{cpp}{std::range::end}

>     \item Ranges concepts are located in \mintinline{cpp}{std::ranges} namespace

Reviewer 2: ```suggestion
    \item Range concepts are located in \mintinline{cpp}{std::ranges} namespace
```

    \item All STL containers are ranges
    \end{itemize}
  \end{block}
  \begin{exampleblock}{Variations on ranges}
    There are actually different type of ranges
    \begin{itemize}

>     \item \mintinline{cpp}{input_range}, \mintinline{cpp}{output_range}, \mintinline{cpp}{bidirectional_range}, \mintinline{cpp}{random_access_range}, \mintinline{cpp}{contiguous_range}, \mintinline{cpp}{forward_range}

Reviewer 2: ```suggestion
    \item \mintinline{cpp}{input_range}, \mintinline{cpp}{output_range}, \mintinline{cpp}{forward_range}, \mintinline{cpp}{bidirectional_range}, \mintinline{cpp}{random_access_range}, \mintinline{cpp}{contiguous_range}
```
Let's order them like the iterators.

>     \item mapping the different types of iterators

Reviewer: I don't know if this is better:
```suggestion
    \item corresponding to the different iterator categories
```

    \end{itemize}
  \end{exampleblock}
\end{frame}

\begin{frame}[fragile]
  \frametitlecpp[20]{Views}
  \begin{block}{The view concept}
    \begin{itemize}

>     \item A \mintinline{cpp}{view} is a non owning object build from a range or another view after applying some range adaptor(s)

Reviewer 2: ```suggestion
    \item A \mintinline{cpp}{view} is a non-owning object built from a range or another view after applying some range adaptor(s)
```

>     \item Its time to copy, move, assign is constant

Reviewer: ```suggestion
    \item The time complexity to copy, move, assign is constant
```

>     \item Range adaptors can be applied with \mintinline{cpp}{|} and are lazy

Reviewer 2: ```suggestion
    \item Range adaptors can be applied/changed with \mintinline{cpp}{|} and are lazy
```

    \end{description}
  \end{block}
  Remember you can combine them via (\mintinline{cpp}{|})
\end{frame}

\begin{frame}[fragile]
  \frametitlecpp[23]{Lazyness and functional programming}
  \begin{exampleblock}{\texttt{view}s are lazy}
    Computation is only trigerred when iterating
    { \scriptsize
      \begin{cppcode*}{gobble=4}
        // print first 20 prime numbers above 1000000
        for (int i: std::views::iota(1000000) | std::views::filter(odd) 
                                              | std::views::filter(isPrime) 
                                              | std::views::take(20)) {
          std::cout << i << " ";  
        }
      \end{cppcode*}
    }
  \end{exampleblock}  
  \begin{block}{\texttt{view}s allow functional programming with the STL}
    { \scriptsize
      \begin{cppcode*}{gobble=4}
        auto const ints = std::views::iota(0, 6);
        auto even = [](int i) { return 0 == i % 2; };
        auto square = [](int i) { return i * i; };
        auto results = numbers | std::views::filter(even)
                               | std::views::transform(square);
        for (auto a : results) { ... } // 0, 4, 16
      \end{cppcode*}
     }

>   \end{block}

Reviewer: Why is this box identical to the example on the previous slide? Did you have in mind to mention something specific about this code?

\subsection[range]{Ranges}

\begin{frame}[fragile]
  \frametitlecpp[20]{Ranges}
  \begin{block}{The range concept}
    \begin{itemize}
    \item A range is a group of items that you can iterator over

>     \item It provides \mintinline{cpp}{begin} and \mintinline{cpp}{end} functions

Reviewer 2: Hmm, it does not need to provide `begin` and `end` in the sense of member functions. A range `r` needs to have `std::ranges::begin(r)` "work" and vice-versa for `end`. I wonder whether we should rephrase the sentence to something like:

```suggestion
    \item It must provide iterators to its beginning and end, obtainable via \mintinline{cpp}{std::range::begin} and \mintinline{cpp}{std::range::end}
```

    \item Ranges concepts are located in \mintinline{cpp}{std::ranges} namespace
    \item All STL containers are ranges
    \end{itemize}
  \end{block}
  \begin{exampleblock}{Variations on ranges}
    There are actually different type of ranges
    \begin{itemize}

>     \item \mintinline{cpp}{input_range}, \mintinline{cpp}{output_range}, \mintinline{cpp}{bidirectional_range},  \mintinline{cpp}{random_access_range},  \mintinline{cpp}{contiguous_range}

Reviewer 2: if you want to be complete you can add `forward_range`

    \item mapping the different types of iterators
    \end{itemize}
  \end{exampleblock}
\end{frame}

\begin{frame}[fragile]
  \frametitlecpp[20]{Views}
  \begin{block}{The view concept}
    \begin{itemize}

>     \item A \mintinline{cpp}{view} is something that you apply on a range and performs some operation

Reviewer 2: What you describe here is actually a range adaptor. But the definition of a view is actually nebulous. See: https://en.cppreference.com/w/cpp/ranges/view
If I come across a better definition I will let you know! :)

On the other hand, range adapters create views.

    \item A \mintinline{cpp}{view} does not own data
    \item It's time to copy, move, assign is constant
    \item \mintinline{cpp}{view}s can be applied with \mintinline{cpp}{|} and are lazy
    \item They live in the \mintinline{cpp}{std::views} namespace
    \end{itemize}
  \end{block}
  \begin{exampleblock}{Example}
    { \scriptsize
      \begin{cppcode*}{gobble=4}

>         auto const ints = {0,1,2,3,4,5};

Reviewer 2: ```suggestion
        auto const ints = std::views::iota(1, 6);
```

        auto even = [](int i) { return 0 == i % 2; };
        auto square = [](int i) { return i * i; };
        auto results = numbers | std::views::filter(even)
                               | std::views::transform(square);
        for (auto a : results) { ... } // 0, 4, 16
      \end{cppcode*}
    }
  \end{exampleblock}
\end{frame}

\begin{frame}[fragile]

>   \frametitlecpp[20]{Many \texttt{view}s are provided by the STL}

Reviewer 2: Again, these are range adapters resulting in the corresponding views. E.g. `std::views::drop` is a range adapter, creating the `std::rangers::drop_view` view.
